### PR TITLE
test: Add Test Cases for Inventory Management Doctypes

### PR DESCRIPTION
**Pull request by rahulsharma8848**
### Title:
Add Test Cases for Inventory Management Doctypes

### Test Summary:
This comprehensive test suite adds automated unit tests for key inventory management doctypes in ERPNext including Inventory Dimension, Bin, Pick List, Purchase Receipt, and Stock Reconciliation. The tests validate core functionality, document workflows, integration points, and edge case handling across inventory operations.

### Doctypes Covered:

Inventory Dimension

Bin

Pick List

Purchase Receipt

Stock Reconciliation

### Scenarios Covered
Inventory Dimension
test_get_parent_fields_TC_SCK_447
Validates correct identification of parent fields referencing inventory dimensions from child doctypes.

Bin
test_update_reserved_qty_for_sub_assembly_TC_SCK_466
Tests reserved quantity updates for production planning with BOM items.

test_on_doctype_update_adds_unique_constraint_TC_SCK_467
Verifies proper addition of unique constraints during doctype updates.

Pick List
test_update_stock_entry_items_with_no_reference_TC_SCK_462
Validates stock entry generation from pick lists without reference documents.

test_update_stock_entry_based_on_work_order_TC_SCK_463
Tests work order integration and proper stock entry creation.

test_get_item_details_TC_SCK_464
Ensures correct item detail retrieval including UOM conversions.

test_update_picked_item_from_current_pick_list_TC_SCK_465
Verifies accurate tracking of picked items with serial/batch handling.

Purchase Receipt
test_get_already_received_qty_TC_SCK_455
Validates calculation of received quantities against purchase orders.

test_check_next_docstatus_TC_SCK_456
Tests document workflow validation with linked purchase invoices.

test_make_item_gl_entries_creates_valid_gl_entries_TC_SCK_457
Verifies proper GL entry generation for inventory transactions.

test_is_landed_cost_booked_for_any_item_TC_SCK_458
Tests landed cost tracking and validation logic.

test_get_billed_qty_against_purchase_receipt_TC_SCK_459
Ensures accurate reconciliation of billed quantities.

test_get_wbs_amount_TC_SCK_WBS_TC_SCK_460
Validates WBS (Work Breakdown Structure) amount calculations.

test_validate_available_budget_paths_TC_SCK_461
Tests all budget validation scenarios including stop/warn conditions.

Stock Reconciliation
test_has_change_in_serial_batch_detects_difference_TC_SCK_486
Validates serial/batch change detection during reconciliation.

test_set_new_serial_and_batch_bundle_logic_TC_SCK_487
Tests proper bundle creation and valuation rate handling.

test_make_adjustment_entry_creates_sle_on_difference_TC_SCK_488
Verifies stock ledger entry generation for quantity differences.

test_make_adjustment_entry_skips_on_zero_difference_TC_SCK_489
Ensures no unnecessary entries are created for zero differences.

test_update_valuation_rate_for_serial_no_applies_to_all_serials_TC_SCK_490
Tests bulk serial number valuation updates.

test_merge_similar_item_serial_nos_merges_correctly_TC_SCK_492
Validates proper merging logic for serial/batch items.

test_has_negative_stock_allowed_behavior_TC_SCK_495
Tests all negative stock validation scenarios.

test_get_difference_account_returns_correct_account_TC_SCK_497
Ensures correct difference account resolution.

### Technology
Framework: unittest using FrappeTestCase base class
Mocks: unittest.mock.patch for dimension/budget testing
Database: Temporary test records via Frappe ORM
Special Configs: Stock Settings overrides for negative stock tests

### Linked Features/Bugs:

N/A (Comprehensive coverage for core inventory functionality)

This structure maintains consistency with your requested format while:

Clearly grouping by doctype

Including all test cases with their validation focus

Adding the requested "Doctypes Covered" section

Keeping technical details in the Technology section

Maintaining clear scenario descriptions

### CODE REVIEW

File: erpnext/stock/doctype/pick_list/test_pick_list.py [hunk 1]
```diff
@@ -1459,48 +1464,284 @@ def test_sales_order_to_sales_invoice_with_2_SI_TC_S_088(self):
...
+		# Setup item and BOM
+		item_code = make_item(
+			"_Test FG Item WO", {"is_stock_item": 1, "stock_uom": "Nos", "valuation_rate": 100}
+		).name
+
+		raw_item = make_item(
+			"_Test Raw Item WO", {"is_stock_item": 1, "stock_uom": "Nos", "valuation_rate": 50}
+		).name
+		# Create required warehouses
+		wip_w = create_warehouse("_Test WIP Warehouse", {"company": "_Test Company"})
```
Comment (aasif-patel) [inline erpnext/stock/doctype/pick_list/test_pick_list.py]: send property is_group as dict and the company last parameter, just check the function definition

**Review by aasif-patel — CHANGES_REQUESTED** [review]

File: erpnext/stock/doctype/bin/test_bin.py [hunk 2]
```diff
@@ -22,14 +26,68 @@ def test_concurrent_inserts(self):
...
+
+		# Create items
+		make_item(raw_material_code, {"is_stock_item": 1})
+		make_item(item_code, {"is_sub_contracted_item": 0, "is_stock_item": 1})
+
+		# Create BOM
+		bom = make_bom(
+			item=item_code, raw_materials=[raw_material_code], rm_qty=2, rate=100, source_warehouse=warehouse
+		)
+		frappe.db.set_value("Item", item_code, "default_bom", bom.name)
```
Comment (aasif-patel) [inline erpnext/stock/doctype/bin/test_bin.py]: make_item returns the item object then why are we using set_value

File: erpnext/stock/doctype/bin/test_bin.py [hunk 3]
```diff
@@ -22,14 +26,68 @@ def test_concurrent_inserts(self):
...
+		bin.reload()
+		self.assertGreaterEqual(bin.reserved_qty_for_production_plan, 0)
+		self.assertIsNotNone(bin.projected_qty)
+
+	def test_on_doctype_update_adds_unique_constraint_TC_SCK_467(self):
+		# Run the hook
+		on_doctype_update()
+
+		# Check if the unique index exists in PostgreSQL
+		indexes = frappe.db.sql("SELECT indexname FROM pg_indexes WHERE tablename = 'tabBin'", as_dict=True)
```
Comment (aasif-patel) [inline erpnext/stock/doctype/bin/test_bin.py]: postgres specific query, it will throw error in mysql envi

File: erpnext/stock/doctype/purchase_receipt/test_purchase_receipt.py [hunk 4]
```diff
@@ -5990,54 +6190,656 @@ def test_get_invoiced_qty_map_TC_SCK_268(self):
...
+		parent_account = ensure_parent_account("Parent Stock Account", company.name, company.abbr)
+		w_account = create_account(
+			account_name="Sub Stock Account",
+			parent_account=parent_account,
+			company=company.name,
+			account_type="Stock",
+			account_currency="INR",
+		)
+
+		frappe.db.set_value("Warehouse", warehouse, "account", w_account)
```
Comment (aasif-patel) [inline erpnext/stock/doctype/purchase_receipt/test_purchase_receipt.py]: pass this property in the create_warehouse method just check how this method creating record

File: erpnext/stock/doctype/purchase_receipt/test_purchase_receipt.py [hunk 5]
```diff
@@ -5990,54 +6190,656 @@ def test_get_invoiced_qty_map_TC_SCK_268(self):
...
+
+		# Explicitly link PI to PR
+		for item in pi.items:
+			item.purchase_receipt = pr.name
+			item.pr_detail = pr.items[0].name
+		pi.save().submit()
+
+		# This should now raise ValidationError due to linked submitted PI
+		with self.assertRaises(frappe.ValidationError, msg="already submitted"):
+			pr.check_next_docstatus()
```
Comment (aasif-patel) [inline erpnext/stock/doctype/purchase_receipt/test_purchase_receipt.py]: use assertIn to check exact same validation msg assertRaises passing all the validations

**Review by aasif-patel — CHANGES_REQUESTED** [review]
Note: Just the functionality or how does the method works when you use it and try to avoid using DB set or db exists id the method already handling it properly

**Comment by aasif-patel** [comment]
and avoid sending One PR for multiple doctype test cases

File: erpnext/stock/doctype/pick_list/test_pick_list.py (hunk 1 above)
Comment (rahulsharma8848) [reply erpnext/stock/doctype/pick_list/test_pick_list.py]: Updated

File: erpnext/stock/doctype/bin/test_bin.py (hunk 2 above)
Comment (rahulsharma8848) [reply erpnext/stock/doctype/bin/test_bin.py]: updated.

File: erpnext/stock/doctype/bin/test_bin.py (hunk 3 above)
Comment (rahulsharma8848) [reply erpnext/stock/doctype/bin/test_bin.py]: updated.

File: erpnext/stock/doctype/purchase_receipt/test_purchase_receipt.py (hunk 4 above)
Comment (rahulsharma8848) [reply erpnext/stock/doctype/purchase_receipt/test_purchase_receipt.py]: updated.

File: erpnext/stock/doctype/bin/test_bin.py [hunk 6]
```diff
@@ -46,13 +46,17 @@ def test_update_reserved_qty_for_sub_assembly_TC_SCK_466(self):
...
 		# Create BOM
 		bom = make_bom(
 			item=item_code, raw_materials=[raw_material_code], rm_qty=2, rate=100, source_warehouse=warehouse
 		)
-		frappe.db.set_value("Item", item_code, "default_bom", bom.name)
+
+		# Again fetching item doc beacuse it refresh after making BOM
+		bom_item = frappe.get_doc("Item", bom_item.name)  # Reload latest version
+		bom_item.default_bom = bom.name
+		bom_item.save()
```
Comment (aasif-patel) [inline erpnext/stock/doctype/bin/test_bin.py]: @rahulsharma8848 we have bom_item as a object we can simply do `bom_item.default_bom = bom.name`

**Review by aasif-patel — CHANGES_REQUESTED** [review]

File: erpnext/stock/doctype/bin/test_bin.py (hunk 6 above)
Comment (rahulsharma8848) [reply erpnext/stock/doctype/bin/test_bin.py]: Updated.

**Review by aasif-patel — APPROVED** [review]

**Comment by swathibollineni2** [comment]
@rahulsharma8848 One test case failed created a new issue https://github.com/8848digital/erpnext/issues/2740
